### PR TITLE
Move doc comment parsing to rustc_lexer

### PR DESCRIPTION
**Pull request by matklad**
Plain comments are trivia, while doc comments are not, so it feels
like this belongs to the rustc_lexer.

The specific reason to do this is the desire to use rustc_lexer in
rustdoc for syntax highlighting, without duplicating "is this a doc
comment?" logic there.

r? @ghost

### CODE REVIEW

File: src/librustc_parse/lexer/mod.rs [hunk 1]
```diff
@@ -318,6 +315,36 @@ impl<'a> StringReader<'a> {
...
+            };
+            self.err_span_(
+                content_start + BytePos(idx as u32 - 1),
+                content_start + BytePos(idx as u32),
+                match comment_kind {
+                    CommentKind::Line => "bare CR not allowed in doc-comment",
+                    CommentKind::Block => "bare CR not allowed in block doc-comment",
+                },
+            );
+        }
```
Comment (matklad) [inline src/librustc_parse/lexer/mod.rs:339]: This loop makes me sad, but zero-cost abstraction are not so zero cost here: https://godbolt.org/z/vE7xsr :(

Comment (Mark-Simulacrum) [reply src/librustc_parse/lexer/mod.rs:339]: `char_indices`-based slow seems to at least unroll somewhat and avoids any panic-handling: https://godbolt.org/z/WP5YaP

Comment (matklad) [reply src/librustc_parse/lexer/mod.rs:339]: Hm, actually, we should just fast-path this...

File: src/librustc_ast/util/comments.rs [hunk 2]
```diff
@@ -216,8 +176,8 @@ pub fn gather_comments(sm: &SourceMap, path: FileName, src: String) -> Vec<Comme
                     }
                 }
             }
-            rustc_lexer::TokenKind::BlockComment { terminated } => {
-                if block_doc_comment_style(token_text, terminated).is_none() {
+            rustc_lexer::TokenKind::BlockComment { kind, terminated: _ } => {
```
Comment (petrochenkov) [inline src/librustc_ast/util/comments.rs:179]:
```suggestion
            rustc_lexer::TokenKind::BlockComment { kind, .. } => {
```

File: src/librustc_ast/util/comments.rs [hunk 3]
```diff
@@ -216,8 +176,8 @@ pub fn gather_comments(sm: &SourceMap, path: FileName, src: String) -> Vec<Comme
...
+                if kind == rustc_lexer::CommentKind::Plain {
```
Comment (petrochenkov) [inline src/librustc_ast/util/comments.rs:180]: Nit: could you `use rustc_lexer::CommentKind` instead?
Long paths like this are not normally used in rustc.

File: src/librustc_parse/lexer/mod.rs [hunk 4]
```diff
@@ -318,6 +315,33 @@ impl<'a> StringReader<'a> {
             }
         }
     }
```
Comment (petrochenkov) [inline src/librustc_parse/lexer/mod.rs:317]: Nit: missing newline between functions.

File: src/librustc_parse/lexer/mod.rs [hunk 5]
```diff
@@ -318,6 +315,33 @@ impl<'a> StringReader<'a> {
...
+                        CommentKind::Line => "bare CR not allowed in doc-comment",
+                        CommentKind::Block => "bare CR not allowed in block doc-comment",
+                    },
+                );
+            }
+        }
+
+        let attr_style = match doc_kind {
+            rustc_lexer::DocCommentKind::Inner => AttrStyle::Inner,
+            rustc_lexer::DocCommentKind::Outer => AttrStyle::Outer,
```
Comment (petrochenkov) [inline src/librustc_parse/lexer/mod.rs:340]: Same comment about `use rustc_lexer::DocCommentKind`.

**Comment by petrochenkov** [comment]
r=me with nits addressed.

File: src/librustc_ast/util/comments.rs (hunk 3 above)
Comment (matklad) [reply src/librustc_ast/util/comments.rs:180]: Ack. I'll leave `rustc_lexer::TokenKind` paths in though, as there are two `TokenKind`s (and this is unrelated to PR)

Comment (matklad) [reply src/librustc_ast/util/comments.rs:180]: Hm, actually, let me take this back. In the `lexer/mod.rs`, if I do replacement, I get

```
    fn cook_doc_comment(
        &self,
        content_start: BytePos,
        content: &str,
        comment_kind: CommentKind,
        doc_kind: DocCommentKind,
    ) -> TokenKind {
```

The problem here is that `CommentKind` and `DocCommentKind` are from different crates, which are confusing.

**Comment by matklad** [comment]
I've addressed nits *except* for qualified paths. I think, *in this case*, qualified paths would be better. 

We have parallel hierarchies of types here:

rustc_lexer: `Token`, `TokenKind`, `CommentKind`, `DocCommentKind`
rustc_ast: `Token`, `TokenKind`, `CommentKind`

So, it seems better if we stick to not qualifying `_ast` and qualifing `_lexer`

File: src/librustc_ast/util/comments.rs (hunk 3 above)
Comment (petrochenkov) [reply src/librustc_ast/util/comments.rs:180]: With this PR `CommentKind` from `rustc_ast` can be eliminated.

Comment (petrochenkov) [reply src/librustc_ast/util/comments.rs:180]: In case of duplicated types `rust_lexer` version have to be qualified, yes.

Comment (matklad) [reply src/librustc_ast/util/comments.rs:180]: Hm, either I disagree, or I don't understand. The two `CommentKind`s are orthogonal axis: one is Inner/Outer, the other is Line/Block. ast's CommentKind (Line/Block) fits the current shape of the ast, it doesn't seem worth it to change it.

What can be merged is `rustc_lexer::CommentKind` and `rustc_ast::AttrStyle`, as they are both just `InnerOrOuter`. I think it's nicer than they are separate though, as they exist on different levels of abstraction, and `AttrStyle` is also serializable.

Comment (matklad) [reply src/librustc_ast/util/comments.rs:180]: Edited https://github.com/rust-lang/rust/pull/75642#issuecomment-676058846 to better show type correspondence. *Independently*, the naming is good :-(

Comment (petrochenkov) [reply src/librustc_ast/util/comments.rs:180]: Ok, I didn't have the code in front of my eyes and mixed everything up.

**Comment by petrochenkov** [comment]
```
CommentKind -----------
----------- CommentKind
```
This looks pretty bad.
I think we need to come up with a different name for one of these.

**Comment by petrochenkov** [comment]
Block vs Line looks like a "heavier" candidate for `CommentKind` to me.

The other can be `CommentIsDoc::(Yes,No)`.

Also `DocCommentKind` -> `DocCommentStyle` (and change its variant order) to mirror `AttrStyle` closer.
I'd probably just put `AttrStyle` as is into the lexer, personally, but there are issues with (en,de)coding and hash impls.

**Comment by matklad** [comment]
How this version look? I've get rid of `CommentKind` and renamed `DocCommentKind -> DocStyle`

**Comment by petrochenkov** [comment]
LGTM.
@bors r+

**Pull request merged**